File: apps/pyapi/api/_rfm.py
"""RFM customer segmentation (pandas).

Score every entity (customer) on Recency, Frequency, Monetary into 1-5 quintiles, then bucket into
recognizable segments (Champions, Loyal, At Risk, …). Fires on transaction-shaped data with an entity id,
a date, and a value column. Pure.
"""
from __future__ import annotations

import re

import numpy as np
import pandas as pd

MIN_CUSTOMERS = 8

# Ordered, mutually-exclusive rules (first match wins); r/f are 1-5 quintile scores.
SEGMENTS = [
    ("champions", "Champions", lambda r, f: r >= 4 and f >= 4),
    ("loyal", "Loyal", lambda r, f: f >= 4 and r >= 2),
    ("potential", "Potential / New", lambda r, f: r >= 4 and f < 4),
    ("at-risk", "At Risk", lambda r, f: r <= 2 and f >= 3),
    ("hibernating", "Hibernating / Lost", lambda r, f: r <= 2 and f < 3),
    ("attention", "Needs Attention", lambda r, f: True),
]

_ENTITY_HINT = re.compile(r"(customer|client|user|account|member|player|buyer|email|cust|patient|subscriber)", re.I)


def _quintile(values: np.ndarray, higher_is_better: bool) -> np.ndarray:
    n = len(values)
    order = np.argsort(values, kind="stable")
    ranks = np.empty(n)
    ranks[order] = np.arange(n)
    score = np.minimum(5, (ranks / n * 5).astype(int) + 1)
    return score if higher_is_better else 6 - score


def _pick(df, profiles):
    entity = next((p for p in profiles if p["role"] == "identifier" and MIN_CUSTOMERS <= p["distinct"] < len(df)), None) \
        or next((p for p in profiles if _ENTITY_HINT.search(p["name"]) and MIN_CUSTOMERS <= p["distinct"] < len(df)), None)
    date = next((p for p in profiles if p["role"] == "time"), None)
    value = next((p for p in profiles if p["role"] == "metric" and p.get("type") == "currency" and p.get("numeric", {}).get("sum", 0) > 0), None) \
        or next((p for p in profiles if p["role"] == "metric" and p.get("numeric", {}).get("sum", 0) > 0), None)
    return entity, date, value


def analyze_rfm(df: pd.DataFrame, profiles: list[dict]) -> dict | None:
    entity, date, value = _pick(df, profiles)
    if not (entity and date and value):
        return None
    work = pd.DataFrame({
        "id": df[entity["name"]].astype("string"),
        "d": pd.to_datetime(df[date["name"]], errors="coerce", format="mixed"),
        "v": pd.to_numeric(df[value["name"]], errors="coerce").clip(lower=0),
    }).dropna(subset=["id", "d"])
    if work.empty:
        return None
    as_of = work["d"].max()
    g = work.groupby("id").agg(last=("d", "max"), freq=("d", "size"), monetary=("v", "sum"))
    if len(g) < MIN_CUSTOMERS:
        return None

    recency = (as_of - g["last"]).dt.days.to_numpy()
    r_score = _quintile(recency, higher_is_better=False)
    f_score = _quintile(g["freq"].to_numpy(), higher_is_better=True)
    keys = [next(k for k, _, m in SEGMENTS if m(int(r), int(f))) for r, f in zip(r_score, f_score)]
    g = g.assign(seg=keys)

    grand = float(g["monetary"].sum()) or 1.0
    segments = []
    for key, label, _ in SEGMENTS:
        rows = g[g["seg"] == key]
        if rows.empty:
            continue
        segments.append({
            "key": key, "label": label, "size": int(len(rows)),
            "sharePct": float(len(rows) / len(g) * 100),
            "avgRecencyDays": float((as_of - rows["last"]).dt.days.mean()),
            "avgFrequency": float(rows["freq"].mean()),
            "avgMonetary": float(rows["monetary"].mean()),
            "totalMonetary": float(rows["monetary"].sum()),
            "monetaryShare": float(rows["monetary"].sum() / grand),
        })
    segments.sort(key=lambda s: s["totalMonetary"], reverse=True)
    return {
        "entity": entity["name"], "dateColumn": date["name"], "valueColumn": value["name"],
        "asOf": as_of.strftime("%Y-%m-%d"), "customers": int(len(g)), "segments": segments,
    }

File: apps/web/src/lib/rfm.ts
import type { ColumnProfile, RfmAnalysis, RfmMember, RfmSegment, Table } from "./types";
import { numericColumn, dateColumn } from "./profile";
import { looksLikeEntity } from "./cohort";

// RFM segmentation - the classic customer-value model. Score every entity (customer) on three axes:
//   Recency   - how long since their last transaction (lower = better),
//   Frequency - how many transactions they made,
//   Monetary  - how much they spent in total,
// each into 1–5 quintiles, then bucket them into recognizable segments (Champions, Loyal, At Risk, …).
// Fires only on transaction-shaped data (an entity id + a date + a value column). Pure + worker-safe.

const MIN_CUSTOMERS = 8;
const DAY = 86_400_000;

interface SegmentDef {
  key: string;
  label: string;
  blurb: string;
  /** first matching rule wins; r/f are 1–5 quintile scores. */
  match: (r: number, f: number) => boolean;
}

// Ordered, mutually-exclusive rules (first match wins) - every (R,F) in 1–5 lands in exactly one bucket.
const SEGMENTS: SegmentDef[] = [
  { key: "champions", label: "Champions", blurb: "Recent, frequent, and high-spending - your best customers.", match: (r, f) => r >= 4 && f >= 4 },
  { key: "loyal", label: "Loyal", blurb: "Buy often and are still active - reward and retain them.", match: (r, f) => f >= 4 && r >= 2 },
  { key: "potential", label: "Potential / New", blurb: "Recent buyers who haven't built a habit yet - nurture them.", match: (r, f) => r >= 4 && f < 4 },
  { key: "at-risk", label: "At Risk", blurb: "Were valuable but haven't bought in a while - win them back.", match: (r, f) => r <= 2 && f >= 3 },
  { key: "hibernating", label: "Hibernating / Lost", blurb: "Low recency and low frequency - largely dormant.", match: (r, f) => r <= 2 && f < 3 },
  { key: "attention", label: "Needs Attention", blurb: "Middling on every axis - a nudge could tip them either way.", match: () => true },
];

/** Assign a 1–5 quintile score to each value by rank position. `higherIsBetter=false` inverts (for recency). */
function quintiles(values: number[], higherIsBetter: boolean): number[] {
  const n = values.length;
  const order = values.map((v, i) => [v, i] as [number, number]).sort((a, b) => a[0] - b[0]);
  const score = new Array<number>(n).fill(3);
  order.forEach(([, origIdx], rank) => {
    let s = Math.min(5, Math.floor((rank / n) * 5) + 1);
    if (!higherIsBetter) s = 6 - s;
    score[origIdx] = s;
  });
  return score;
}

function pickColumns(table: Table, profiles: ColumnProfile[]): { entity: ColumnProfile; date: ColumnProfile; value: ColumnProfile } | null {
  // A repeating customer column is usually typed "category" (not "id"), so - like cohort analysis - fall
  // back to an id-ish-named column that recurs. `distinctCount < rowCount` ensures real repeat custom.
  const entity =
    profiles.find((p) => p.role === "identifier" && p.distinctCount >= MIN_CUSTOMERS && p.distinctCount < table.rowCount) ??
    profiles.find((p) => looksLikeEntity(p.name) && p.distinctCount >= MIN_CUSTOMERS && p.distinctCount < table.rowCount);
  const date = profiles.find((p) => p.role === "time" && p.type === "date");
  // Prefer a currency column for monetary; fall back to the largest-sum metric.
  const value =
    profiles.find((p) => p.role === "metric" && p.type === "currency" && (p.numeric?.sum ?? 0) > 0) ??
    profiles
      .filter((p) => p.role === "metric" && p.numeric && (p.numeric.sum ?? 0) > 0)
      .sort((a, b) => (b.numeric!.sum ?? 0) - (a.numeric!.sum ?? 0))[0];
  if (!entity || !date || !value) return null;
  return { entity, date, value };
}

interface RfmDetail {
  entity: string;
  dateColumn: string;
  valueColumn: string;
  asOfDay: number;
  members: RfmMember[];
}

/** Score every entity into RFM members - the shared core behind both the summary and the export. */
function rfmDetail(table: Table, profiles: ColumnProfile[]): RfmDetail | undefined {
  const cols = pickColumns(table, profiles);
  if (!cols) return undefined;

  const dates = dateColumn(table, cols.date.name);
  const values = numericColumn(table, cols.value.name);

  // Roll transactions up per entity: last-seen day, transaction count, total spend.
  interface Acc { lastDay: number; freq: number; monetary: number }
  const byEntity = new Map<string, Acc>();
  let asOfDay = -Infinity;
  table.rows.forEach((r, i) => {
    const iso = dates[i];
    if (!iso) return; // a transaction needs a date to score recency
    const day = Math.floor(Date.parse(iso) / DAY);
    if (!Number.isFinite(day)) return;
    const id = r[cols.entity.name];
    if (id === null || id === undefined || id === "") return;
    const key = String(id);
    if (day > asOfDay) asOfDay = day;
    const acc = byEntity.get(key) ?? { lastDay: -Infinity, freq: 0, monetary: 0 };
    acc.freq += 1;
    if (day > acc.lastDay) acc.lastDay = day;
    const v = values[i];
    if (Number.isFinite(v) && v >= 0) acc.monetary += v;
    byEntity.set(key, acc);
  });

  const ids = [...byEntity.keys()];
  if (ids.length < MIN_CUSTOMERS || !Number.isFinite(asOfDay)) return undefined;

  const recencyDays = ids.map((id) => asOfDay - byEntity.get(id)!.lastDay);
  const freq = ids.map((id) => byEntity.get(id)!.freq);
  const monetary = ids.map((id) => byEntity.get(id)!.monetary);

  const rScore = quintiles(recencyDays, false); // fewer days since last seen = better
  const fScore = quintiles(freq, true);

  const members: RfmMember[] = ids.map((id, i) => {
    const def = SEGMENTS.find((s) => s.match(rScore[i], fScore[i]))!;
    return {
      id,
      recencyDays: recencyDays[i],
      frequency: freq[i],
      monetary: monetary[i],
      rScore: rScore[i],
      fScore: fScore[i],
      segmentKey: def.key,
      segmentLabel: def.label,
    };
  });

  return { entity: cols.entity.name, dateColumn: cols.date.name, valueColumn: cols.value.name, asOfDay, members };
}

/** The per-entity membership behind the RFM segments - used to export a segment as a worklist. */
export function rfmMembers(table: Table, profiles: ColumnProfile[]): RfmMember[] | undefined {
  return rfmDetail(table, profiles)?.members;
}

export function analyzeRfm(table: Table, profiles: ColumnProfile[]): RfmAnalysis | undefined {
  const detail = rfmDetail(table, profiles);
  if (!detail) return undefined;
  const { members } = detail;

  // Aggregate entities into their segments.
  interface Bucket { size: number; recency: number; freq: number; monetary: number }
  const buckets = new Map<string, Bucket>();
  for (const m of members) {
    const b = buckets.get(m.segmentKey) ?? { size: 0, recency: 0, freq: 0, monetary: 0 };
    b.size += 1;
    b.recency += m.recencyDays;
    b.freq += m.frequency;
    b.monetary += m.monetary;
    buckets.set(m.segmentKey, b);
  }

  const grandMonetary = members.reduce((s, m) => s + m.monetary, 0) || 1;
  const segments: RfmSegment[] = SEGMENTS.filter((d) => buckets.has(d.key)).map((d) => {
    const b = buckets.get(d.key)!;
    return {
      key: d.key,
      label: d.label,
      blurb: d.blurb,
      size: b.size,
      sharePct: (b.size / members.length) * 100,
      avgRecencyDays: b.recency / b.size,
      avgFrequency: b.freq / b.size,
      avgMonetary: b.monetary / b.size,
      totalMonetary: b.monetary,
      monetaryShare: b.monetary / grandMonetary,
    };
  });
  segments.sort((a, b) => b.totalMonetary - a.totalMonetary);

  return {
    entity: detail.entity,
    dateColumn: detail.dateColumn,
    valueColumn: detail.valueColumn,
    asOf: new Date(detail.asOfDay * DAY).toISOString().slice(0, 10),
    customers: members.length,
    segments,
  };
}
